File: problems/0006_zigzag-conversion/Solution.py
#!/usr/bin/python3

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        sLen = len(s)
        if sLen <= 2 or numRows == 1:
            return s
        colsPerPad = numRows - 1
        charsPerPad = colsPerPad - 1 + numRows
        lastRow = colsPerPad
        def g():
            for r in range(numRows):
                for i in range(r, sLen, charsPerPad):
                    yield s[i]
                    j = i + charsPerPad - 2 * r
                    if r != 0 and r != lastRow and j < sLen:
                        yield s[j]
        return ''.join(g())

File: problems/0006_zigzag-conversion/solution.go
func convert(s string, numRows int) string {
    sLen := len(s)
    if sLen <= 2 || numRows == 1 {
        return s
    }
    runes, colsPerPad := []rune(s), numRows - 1
    charsPerPad, lastRow := colsPerPad - 1 + numRows, colsPerPad
    result := make([]rune, sLen)
    for r, k := 0, 0; r < numRows; r++ {
        for i := r; i < sLen; i += charsPerPad {
            result[k] = runes[i]
            k++
            if r != 0 && r != lastRow {
                j := i + charsPerPad - 2 * r
                if j < sLen {
                    result[k] = runes[j]
                    k++
                }
            }
        }
    }
    return string(result)
}
